# **Data Wrangling**


## **What Is Data Wrangling?**

Data wrangling (also known as data munging) is the process of converting and formating data from its raw form to usable format further down the data science pipeline.

## **What Is the Need for Data Wrangling?**

Without feeding proper data into a model, one cannot expect a model that is dependable and gives higher accuracy. 

## **Steps involve in  Data Wrangling**


### Data Acquisition:

 Collecting data from various sources, which may include databases, files, APIs, or web scraping.

### Data Cleaning

Identifying and handling missing or incomplete data, correcting errors, and resolving inconsistencies. This step may involve tasks such as imputation, removal of duplicates, and dealing with outliers.

### Data Transformation

Restructuring or converting the data into a more appropriate format for analysis. This can include tasks such as changing data types, reshaping data, or creating new variables.

### Data Enrichment

Enhancing the dataset with additional information from external sources to provide more context or depth for analysis.

### Data Aggregation

Combining multiple datasets or summarizing data to create aggregated views for analysis.

### Data Validation

Ensuring that the processed data meets quality standards and is suitable for analysis.

# Scenario 1

### <b> Problem Statement </b>
 
You are a junior data scientist and you are assigned a new task to perform data wrangling on a set of datasets. The datasets have many ambiguities. You have to identify those and apply different data wrangling techniques to get a dataset for further usage.

#### <b> Dataset </b>

- Download the `rental_bike_descr, rental_bike_season, and final_rental_bike_dataset` from Course Resources and upload the datasets to the lab

#### **Data Dictionary**


**Attribute Information:**


* date = date of the ride  
* season -  1 = spring, 2 = summer, 3 = fall, 4 = winter 
* holiday - whether the day is considered a holiday
* workingday - whether the day is neither a weekend nor holiday
* weather - 1: Clear, Few clouds, Partly cloudy, Partly cloudy
            2: Mist + Cloudy, Mist + Broken clouds, Mist + Few clouds, Mist
            3: Light Snow, Light Rain + Thunderstorm + Scattered clouds, Light Rain + Scattered clouds
            4: Heavy Rain + Ice Pallets + Thunderstorm + Mist, Snow + Fog 
* temp - temperature in Celsius
* atemp - "feels like" temperature in Celsius
* humidity - relative humidity
* windspeed - wind speed
* casual - number of non-registered user rentals initiated
* registered - number of registered user rentals initiated
* count - number of total rentals

### **Import libraries**

- Pandas is a high-level data manipulation tool

- NumPy is used for working with multidimensional arrays

In [1]:
import pandas as pd
import numpy as np

In [2]:

# Displays the versions of various libraries
print(pd.show_versions())

c:\Users\HariharanSivakumar\AppData\Local\Programs\Python\Python311\Lib\site-packages\_distutils_hack\__init__.py:33: UserWarning: Setuptools is replacing distutils.
  warnings.warn("Setuptools is replacing distutils.")



INSTALLED VERSIONS
------------------
commit                : bdc79c146c2e32f2cab629be240f01658cfb6cc2
python                : 3.11.3.final.0
python-bits           : 64
OS                    : Windows
OS-release            : 10
Version               : 10.0.19045
machine               : AMD64
processor             : Intel64 Family 6 Model 140 Stepping 1, GenuineIntel
byteorder             : little
LC_ALL                : None
LANG                  : None
LOCALE                : English_India.1252

pandas                : 2.2.1
numpy                 : 1.26.4
pytz                  : 2024.1
dateutil              : 2.9.0.post0
setuptools            : 65.5.0
pip                   : 22.3.1
Cython                : None
pytest                : None
hypothesis            : None
sphinx                : None
blosc                 : None
feather               : None
xlsxwriter            : None
lxml.etree            : None
html5lib              : None
pymysql               : None
psycopg2         

### **Load the first dataset**

In [3]:
dataset_1 = pd.read_csv('data/rental_bike_descr.csv')

#### **Observations**:
- We have to upload the dataset 
- We are reading the file through the dataset_1 variable

- The file is in CSV format

- We use the **`pd.read_csv()`** function to read a CSV file

- We provide the exact path of the file within the round bracket **`()`**

### **Check the type of dataset**

- Execute the below command to understand type of data we are having

In [4]:
type(dataset_1)

pandas.core.frame.DataFrame

#### **Observations:**

- The result shows that the dataset is DataFrame

- DataFrame is a tabular structure consisting of rows and columns

### **Shape of the dataset**

In [5]:
dataset_1.shape

(610, 10)

#### **Observation**:

- The `dataset_1` has 610 rows and 10 columns.

### **Print first 5 rows of the dataset**

In [6]:
dataset_1.head()

instant      dteday  season  yr  mnth  hr  holiday  weekday  weathersit  \
0        1  01-01-2011       1   0     1   0    False        6           1   
1        2  01-01-2011       1   0     1   1    False        6           1   
2        3  01-01-2011       1   0     1   2    False        6           1   
3        4  01-01-2011       1   0     1   3    False        6           1   
4        5  01-01-2011       1   0     1   4    False        6           1   

   temp  
0  0.24  
1  0.22  
2  0.22  
3  0.24  
4  0.24

#### **Observation**:

- The **'dataset_1.head()'** function displays only the initial five rows of the dataset.

### **Load the second dataset**

- Use the function carefully since it is an excel file

In [7]:
dataset_2 = pd.read_csv('data/rental_bike_season.csv')

### **Shape of the dataset**

In [8]:
dataset_2.shape

(610, 8)

#### **Observation:**

* The result shows that `dataset_2` has 610 rows and 8 columns.

### **Print first 5 rows of the dataset**

In [9]:
dataset_2.head()

Unnamed: 0  instant   atemp   hum  windspeed  casual  registered  cnt
0           0        1  0.2879  0.81        0.0       3          13   16
1           1        2  0.2727  0.80        0.0       8          32   40
2           2        3  0.2727  0.80        0.0       5          27   32
3           3        4  0.2879  0.75        0.0       3          10   13
4           4        5  0.2879  0.75        0.0       0           1    1

#### **Observation:**
- We can see a column named `unnamed:0`, which is not in the data dictionary. Let's remove it.

### **Drop the column**

In [10]:
dataset_2 = dataset_2.drop(['Unnamed: 0'], axis=1)

### **Lets check the shape of the dataset again after the drop**

In [11]:
dataset_2.shape

(610, 7)

#### **Observation:**

- We had 8 columns before the drop.

- When we check the shape of the file after the drop, we see that the column `Unnamed: 0` has been dropped

### **Top 5 rows of the dataset**

- Let's check the dataset_2 again

In [12]:
dataset_2.head()

instant   atemp   hum  windspeed  casual  registered  cnt
0        1  0.2879  0.81        0.0       3          13   16
1        2  0.2727  0.80        0.0       8          32   40
2        3  0.2727  0.80        0.0       5          27   32
3        4  0.2879  0.75        0.0       3          10   13
4        5  0.2879  0.75        0.0       0           1    1

#### **Observation:**

- dataset_2 does not have `Unnamed: 0` column

### **Merge the datasets**

- We have two datasets. They are dataset_1 and dataset_2

- As both datasets have one common column 'instant', let's merge the datasets on that column

- We are going to save the resultant data inside the **combined_data** as shown below

In [13]:
combined_data = pd.merge(dataset_1, dataset_2, on='instant')

### **Check the shape of combined dataset**

In [14]:
combined_data.shape

(610, 16)

#### **Observation:**

- The shape of the combined_data has 610 rows and 16 columns 

### **Top 5 rows of the combined dataset**

In [15]:
combined_data.head()

instant      dteday  season  yr  mnth  hr  holiday  weekday  weathersit  \
0        1  01-01-2011       1   0     1   0    False        6           1   
1        2  01-01-2011       1   0     1   1    False        6           1   
2        3  01-01-2011       1   0     1   2    False        6           1   
3        4  01-01-2011       1   0     1   3    False        6           1   
4        5  01-01-2011       1   0     1   4    False        6           1   

   temp   atemp   hum  windspeed  casual  registered  cnt  
0  0.24  0.2879  0.81        0.0       3          13   16  
1  0.22  0.2727  0.80        0.0       8          32   40  
2  0.22  0.2727  0.80        0.0       5          27   32  
3  0.24  0.2879  0.75        0.0       3          10   13  
4  0.24  0.2879  0.75        0.0       0           1    1

### **Load the third dataset**

#### **Import the dataset**

In [16]:
dataset_3 = pd.read_csv('data/final_rental_bike_dataset.csv')

### **Check the shape of the dataset**

In [17]:
dataset_3.shape

(390, 16)

### **Top 5 rows of the dataset**

In [18]:
dataset_3.head()

instant      dteday  season  yr  mnth  hr  holiday  weekday  weathersit  \
0      620  29-01-2011       1   0     1   1    False        6           1   
1      621  29-01-2011       1   0     1   2    False        6           1   
2      622  29-01-2011       1   0     1   3    False        6           1   
3      623  29-01-2011       1   0     1   4    False        6           1   
4      624  29-01-2011       1   0     1   6    False        6           1   

   temp   atemp   hum  windspeed  casual  registered  cnt  
0  0.22  0.2273  0.64     0.1940       0          20   20  
1  0.22  0.2273  0.64     0.1642       0          15   15  
2  0.20  0.2121  0.64     0.1343       3           5    8  
3  0.16  0.1818  0.69     0.1045       1           2    3  
4  0.16  0.1818  0.64     0.1343       0           2    2

### **Bottom 15 rows of the dataset**

- Just like the `head` function, the `tail` function is used to see the bottom rows of the dataset

- If you want to see the specific number of rows, then specify the number inside the **`bracket ()`** as shown below

In [19]:
dataset_3.tail(15)

instant      dteday  season  yr  mnth  hr  holiday  weekday  weathersit  \
375      995  14-02-2011       1   0     2   2    False        1           1   
376      996  14-02-2011       1   0     2   3    False        1           1   
377      997  14-02-2011       1   0     2   4    False        1           1   
378      998  14-02-2011       1   0     2   5    False        1           1   
379      999  14-02-2011       1   0     2   6    False        1           1   
380     1000  14-02-2011       1   0     2   7    False        1           1   
381      611  28-01-2011       1   0     1  16    False        5           1   
382      612  28-01-2011       1   0     1  17    False        5           1   
383      613  28-01-2011       1   0     1  18    False        5           1   
384      614  28-01-2011       1   0     1  19    False        5           2   
385      615  28-01-2011       1   0     1  20    False        5           2   
386      616  28-01-2011       1   0     1  21    False        5           2   
387      617  28-01-2011       1   0     1  22    False        5           1   
388      618  28-01-2011       1   0     1  23    False        5           1   
389      619  29-01-2011       1   0     1   0    False        6           1   

     temp   atemp   hum  windspeed  casual  registered  cnt  
375  0.36  0.3333  0.40     0.2985       0           2    2  
376  0.34  0.3182  0.46     0.2239       1           1    2  
377  0.32  0.3030  0.53     0.2836       0           2    2  
378  0.32  0.3030  0.53     0.2836       0           3    3  
379  0.34  0.3030  0.46     0.2985       1          25   26  
380  0.34  0.3030  0.46     0.2985       2          96   98  
381  0.22  0.2727  0.80     0.0000      10          70   80  
382  0.24  0.2424  0.75     0.1343       2         147  149  
383  0.24  0.2273  0.75     0.1940       2         107  109  
384  0.24  0.2424  0.75     0.1343       5          84   89  
385  0.24  0.2273  0.70     0.1940       1          61   62  
386  0.22  0.2273  0.75     0.1343       1          57   58  
387  0.24  0.2121  0.65     0.3582       0          26   26  
388  0.24  0.2273  0.60     0.2239       1          22   23  
389  0.22  0.1970  0.64     0.3582       2          26   28

#### **Observation:**

- The bottom 15 rows of the dataset_3 is shown above, as we mention 15 inside the bracket ()

- Here, we can see that the rows are not sorted well according to the `instant` number. Let's resolve it.

### **Sort values of a column**

- To sort the values per our will, we use the **`sort_values`** function and in the square brackets, we specify the name of the column by which we want to sort, as shown below

In [20]:
dataset_3 = dataset_3.sort_values(by=['instant'])

- Let's check head and tail to verify the sort operation

In [21]:
dataset_3.head()

instant      dteday  season  yr  mnth  hr  holiday  weekday  weathersit  \
381      611  28-01-2011       1   0     1  16    False        5           1   
382      612  28-01-2011       1   0     1  17    False        5           1   
383      613  28-01-2011       1   0     1  18    False        5           1   
384      614  28-01-2011       1   0     1  19    False        5           2   
385      615  28-01-2011       1   0     1  20    False        5           2   

     temp   atemp   hum  windspeed  casual  registered  cnt  
381  0.22  0.2727  0.80     0.0000      10          70   80  
382  0.24  0.2424  0.75     0.1343       2         147  149  
383  0.24  0.2273  0.75     0.1940       2         107  109  
384  0.24  0.2424  0.75     0.1343       5          84   89  
385  0.24  0.2273  0.70     0.1940       1          61   62

In [22]:
dataset_3.tail()

instant      dteday  season  yr  mnth  hr  holiday  weekday  weathersit  \
376      996  14-02-2011       1   0     2   3    False        1           1   
377      997  14-02-2011       1   0     2   4    False        1           1   
378      998  14-02-2011       1   0     2   5    False        1           1   
379      999  14-02-2011       1   0     2   6    False        1           1   
380     1000  14-02-2011       1   0     2   7    False        1           1   

     temp   atemp   hum  windspeed  casual  registered  cnt  
376  0.34  0.3182  0.46     0.2239       1           1    2  
377  0.32  0.3030  0.53     0.2836       0           2    2  
378  0.32  0.3030  0.53     0.2836       0           3    3  
379  0.34  0.3030  0.46     0.2985       1          25   26  
380  0.34  0.3030  0.46     0.2985       2          96   98

### **Concatenate the `combine_data` with `dataset_3`**

- Let's concatenate both DataFrame combined_data and dataset_3 into a single DataFrame using the **`concat`** function, as shown below
- Store the final DataFrame inside the `final_data` variable

In [23]:
final_data = pd.concat([combined_data, dataset_3])

### **Check the shape of the new dataset**

In [24]:
final_data.shape

(1000, 16)

#### **Observation:**

- Now, the `final_data` has 1000 rows and 16 columns



### **Let's diplay the columns of the `final_data` DataFrame**

In [25]:
final_data = final_data.rename(columns={'dteday': 'date', 'yr': 'year', 'mnth':'month','hr':'hour',
                                        'weathersit':'weather', 'hum':'humidity', 'cnt':'count'})

In [26]:
final_data.head()

instant        date  season  year  month  hour  holiday  weekday  weather  \
0        1  01-01-2011       1     0      1     0    False        6        1   
1        2  01-01-2011       1     0      1     1    False        6        1   
2        3  01-01-2011       1     0      1     2    False        6        1   
3        4  01-01-2011       1     0      1     3    False        6        1   
4        5  01-01-2011       1     0      1     4    False        6        1   

   temp   atemp  humidity  windspeed  casual  registered  count  
0  0.24  0.2879      0.81        0.0       3          13     16  
1  0.22  0.2727      0.80        0.0       8          32     40  
2  0.22  0.2727      0.80        0.0       5          27     32  
3  0.24  0.2879      0.75        0.0       3          10     13  
4  0.24  0.2879      0.75        0.0       0           1      1

### **Data types of different column values**

In [27]:
final_data.dtypes

instant         int64
date           object
season          int64
year            int64
month           int64
hour            int64
holiday          bool
weekday         int64
weather         int64
temp          float64
atemp         float64
humidity      float64
windspeed     float64
casual          int64
registered      int64
count           int64
dtype: object





#### **Observations:**

- We can see that the majority of our data columns are of type int64. They are therefore 64-bit integers. Some of the columns are of the type float64, which implies that they have decimals in them. However, only the date column has an object type, indicating that it contains strings.

### **Check for null values**

- Execute the given command to check the unknown values in the DataFrame


In [28]:
final_data.isna()

instant   date  season   year  month   hour  holiday  weekday  weather  \
0      False  False   False  False  False  False    False    False    False   
1      False  False   False  False  False  False    False    False    False   
2      False  False   False  False  False  False    False    False    False   
3      False  False   False  False  False  False    False    False    False   
4      False  False   False  False  False  False    False    False    False   
..       ...    ...     ...    ...    ...    ...      ...      ...      ...   
376    False  False   False  False  False  False    False    False    False   
377    False  False   False  False  False  False    False    False    False   
378    False  False   False  False  False  False    False    False    False   
379    False  False   False  False  False  False    False    False    False   
380    False  False   False  False  False  False    False    False    False   

      temp  atemp  humidity  windspeed  casual  registered  count  
0    False  False     False      False   False       False  False  
1    False  False     False      False   False       False  False  
2    False  False     False      False   False       False  False  
3    False  False     False      False   False       False  False  
4    False  False     False      False   False       False  False  
..     ...    ...       ...        ...     ...         ...    ...  
376  False  False     False      False   False       False  False  
377  False  False     False      False   False       False  False  
378  False  False     False      False   False       False  False  
379  False  False     False      False   False       False  False  
380  False  False     False      False   False       False  False  

[1000 rows x 16 columns]

#### **Observation:**

- The **`isna()`** function returns DataFrame of Boolean values that are True for null values  

- In a huge dataset, the code given above is not going to help

- We do not get enough idea of the null values by looking at the given tabular dataset

- The next line of code is more convenient in this case.

In [29]:
final_data.isna().sum(axis=0)

instant        0
date           0
season         0
year           0
month          0
hour           0
holiday        0
weekday        0
weather        0
temp           0
atemp         11
humidity       0
windspeed      0
casual         0
registered     0
count          0
dtype: int64

#### **Observations:**

- The `isna().sum(axis=0)` function provides a clear picture of the number of null values in a DataFrame

-  In the given result, we can see that the `atemp` column has 11 null values

### **Let's check the percentage of the rows with missing value**

- We are performing this operation to determine whether the NA value rows can be dropped off or not so that we cannot deviate from our desired model

In [30]:
percentage_of_missing_values = (final_data['atemp'].isna().sum(axis=0)/final_data.shape[0])*100
percentage_of_missing_values

1.0999999999999999

#### **Observations:**

- We divide the number of null values by the shape of the DataFrame to get the percentage of missing values.

- Since the percentage is 1, it is very less. Usually, the industry practice allows us to drop rows up to 30%. So, we can drop the rows with missing values.

### **Drop the rows with missing values**

- We will use the `dropna` function to drop the null value rows

In [31]:
final_data = final_data.dropna(axis=0)
final_data.shape

(989, 16)

#### **Observations:**

- We can see that the shape of the DataFrame reduced to 989 from 1000. It shows that the missing value rows have been wiped off.

- In further lessons of this course, we'll see different methods to treat missing values.

### **Now, let's again check the missing value count after the drop**

In [32]:
final_data.isna().sum(axis=0)

instant       0
date          0
season        0
year          0
month         0
hour          0
holiday       0
weekday       0
weather       0
temp          0
atemp         0
humidity      0
windspeed     0
casual        0
registered    0
count         0
dtype: int64

#### **Perform sanity checks on the dataset**

- It verifies the logical correctness of the data points

### **Check if `casual` + `registered` is always equal to `count`**

In [33]:
np.sum(final_data['casual'] + final_data['registered'] - final_data['count'])

0

### **Month values should be in the range of 1-12**

- We will use the `unique()` function to find the elements of an array

In [34]:
np.unique(final_data.month)

array([1, 2], dtype=int64)

### **Hour should be in the range of 1-24**

In [35]:
np.unique(final_data.hour)

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23], dtype=int64)

### **Print the statistical summary of the data**

- We will use the `describe()` function to see the stastical summary of the dataset

In [37]:
print(final_data.describe())

           instant  season   year       month        hour     weekday  \
count   989.000000   989.0  989.0  989.000000  989.000000  989.000000   
mean    505.622851     1.0    0.0    1.315470   11.753286    2.991911   
std     286.274765     0.0    0.0    0.464938    6.891129    2.084727   
min       1.000000     1.0    0.0    1.000000    0.000000    0.000000   
25%     259.000000     1.0    0.0    1.000000    6.000000    1.000000   
50%     506.000000     1.0    0.0    1.000000   12.000000    3.000000   
75%     753.000000     1.0    0.0    2.000000   18.000000    5.000000   
max    1000.000000     1.0    0.0    2.000000   23.000000    6.000000   

          weather        temp       atemp    humidity   windspeed      casual  \
count  989.000000  989.000000  989.000000  989.000000  989.000000  989.000000   
mean     1.479272    0.204712    0.211958    0.581769    0.194609    4.921132   
std      0.651085    0.077789    0.076703    0.187706    0.129225    7.666231   
min      1.000000 